### DISCUSSION TITLE
python example plugin setting  is_filter=true is ignored

### DISCUSSION THREAD

**Discussion by shemeshg** [nushell/nushell#5933] [Q&A] [answered]
Add is_filter to signature of 

https://github.com/nushell/nushell/blob/main/crates/nu_plugin_python/plugin.py

```python
def signatures():
    """
    Multiple signatures can be sent to Nushell. Each signature will be registered
    as a different plugin function in Nushell.

    In your plugin logic you can use the name of the signature to indicate what
    operation should be done with the plugin
    """
    return {
        "Signature": [
            {
                "name": "nu-python",
                "usage": "Signature test for Python",
                "extra_usage": "",
                "is_filter": True,
                "required_positional": [
```
- rm plugin.nu
- register --encoding=json ./nu_plugin_gogo
Check settings in plugin.nu
```
% cat plugin.nu|grep is_filter
  "is_filter": false,
%
```

- tried manually change `"is_filter": true,` in `plugin.nu`,  and plugin behaves the same,
as if `"is_filter": false,`

**Comment by fdncred** [comment, marked as answer]
Sounds like a bug but then again, maybe we should just remove it. I'm not sure it even does anything.

**Reply by shemeshg** [reply]
# yes
1. It seems to stopped doing anything since nushell abandoned JsonRpc2.0 standard.
2. all Rust examples are without it also ...
3. Some builtin commands also have this problem
for example: `ls */**|first 5`  will not print single line until iterating the entire drive. 

# My wish is
1. I do like to parse the input stream, and provide output when input size is infinite/undetermined or unknown
2. All filter category internal commands, do work as expected, but I do not understand why and how.
  See all built in filter commands under here 
  https://github.com/nushell/nushell/tree/main/crates/nu-command/src/filters

### Maybe Typescript example how stream might look
```typescript
abstract class StreamIO {
  isOpened = true;

  protected abstract localIsClosed(): boolean;
  outStream?: StreamIO;

  updateIsOpen(): void {
    if (this.localIsClosed()) {
      this.isOpened = false;
    }
    if (this.outStream) {
      this.outStream.updateIsOpen();
      if (!this.outStream.isOpened) { this.isOpened = false }
    }
  }

  process(s: string): string {
    if (this.outStream) {
      return this.outStream.process(this.localProcess(s));
    } else {
      return this.localProcess(s);
    }
  }

  protected abstract localProcess(s: string): string;
}



class Head extends StreamIO {
  counter = 0;
  constructor(i: number) {
    super();
    this.counter = i;
  }

  protected localProcess(s: string): string {
    this.counter--;
    return s;
  }

  protected localIsClosed(): boolean {
    return this.counter === 0;
  }
}

class AddGGGToEachLine extends StreamIO {
  
  protected localProcess(s: string): string {
    return "GGG " + s;
  }

  protected localIsClosed(): boolean {
    return false;
  }
}


class TextGenerator extends StreamIO {
  
  protected localProcess(s: string): string {
    return s;
  }
  protected localIsClosed(): boolean {
    return this.a.length == 0;
  }

  private a: string[] = []
  
  addLine(s: string) {
    this.a.push(s);
  }

  getLine(): string {
    const s = this.a.shift() || ""
    return s;
  }


}

function main(): void {
  let a = new TextGenerator();
  for (let i = 0; i < 10; i++) {
    a.addLine("Hi" + i)
  }
  a.outStream = new Head(3);
  a.outStream.outStream=new AddGGGToEachLine();

  a.updateIsOpen();
  while (a.isOpened) {
    console.log(a.process(a.getLine()))
    a.updateIsOpen();
  }

}

main();

```

will print:
```
GGG Hi0
GGG Hi1
GGG Hi2
```
### another option is passing the stream to subprocesses like in xonsh
xonsh example:
```python
def mycmd3(args, stdin=None, stdout=None):
  for line in stdin:
    print ("whatever " + line.strip(), file=stdout )
```

# My far future wishful thinking
I wish plugins would be just Shared Object dynamically loaded on runtime
```cpp
// #include <of dummy Skelton class each time overrides by next plugin to load>
 void* handle = dlopen("myPluginClass.so", RTLD_LAZY);
```
So I could have static variables

For example, this would enable 
register "./my-command.so"
my-command login_to_db
my-command "select whatever"

because this way It would be possible to maintain state  (static variable) like in PoweShell and CmdLets

(This actually eliminate the need for any Json communication, since the interface functions of what will be using Rust implementation function in the shared object)
